### ISSUE TITLE
Have benchmark actions leave a comment in PRs

### ISSUE THREAD

**Issue by andrewiggins** [preactjs/preact#2576]
Leaving myself a note here that'll be easy to find as I work on improving the benchmark experience:

Per [comment](https://github.com/preactjs/preact/pull/2560#issuecomment-636169654) by @developit in #2560:

This is absolutely stellar.

Two thoughts:
* [x] ~~Do you think it would be better to consolidate the build so that it builds once and uses that same build for all benchmarks?~~ Done.  
* [x] ~~Also, I wonder if we could have it post the benchmark results as a comment?~~ [Done](https://github.com/andrewiggins/tachometer-reporter-action).

For the comment posting, here's some helper functions I have that use actions-toolkit:

```js
/** Create a status check, and return a function that updates (completes) it. */
async function createCheck(github, context) {
    const check = await github.checks.create({
        ...context.repo,
        name: 'Benchmarks',
        head_sha: context.payload.pull_request.head.sha,
        status: 'in_progress',
    });

    return async details => {
        await github.checks.update({
            ...context.repo,
            check_run_id: check.data.id,
            completed_at: new Date().toISOString(),
            status: 'completed',
            ...details
        });
    };
}


/** create a PR comment, or update one if it already exists */
async function postOrUpdateComment(github, context, commentMarkdown) {
    const commentInfo = {
		...context.repo,
		issue_number: context.issue.number
	};

	const comment = {
		...commentInfo,
		body: commentMarkdown + '\n\n<sub>preact-benchmarks-action</sub>' // used to update this comment later
	};

	startGroup(`Updating PR comment`);
	let commentId;
	try {
		const comments = (await github.issues.listComments(commentInfo)).data;
		for (let i=comments.length; i--; ) {
			const c = comments[i];
			if (c.user.type === 'Bot' && /<sub>[\s\n]*preact-benchmarks-action/.test(c.body)) {
				commentId = c.id;
				break;
			}
		}
	}
	catch (e) {
		console.log('Error checking for previous comments: ' + e.message);
	}

	if (commentId) {
		try {
			await github.issues.updateComment({
				...context.repo,
				comment_id: commentId,
				body: comment.body
			});
		}
		catch (e) {
			commentId = null;
		}
	}

	if (!commentId) {
		try {
			await github.issues.createComment(comment);
		} catch (e) {
			console.log(`Error creating comment: ${e.message}`);
        }
    }
    endGroup();
}
```

**Usage:**

```js
// example action.js

import { getInput, startGroup, endGroup, setFailed, setOutput } from '@actions/core';
import { GitHub, context } from '@actions/github';

async function runBenchmarks(github, context) {
  return {
    markdown: 'full-page markdown result',
    summary: 'quick one-line stats summary'
  };
}

(async () => {
    const token = process.env.GITHUB_TOKEN || getInput('repo-token');
    const github = token ? new GitHub(token) : {};

    // if there's no GITHUB_TOKEN, just log to stdout
    let finish = details => console.log(details);
    if (token) {
        finish = await createCheck(github, context);
    }

    try {
        // call the actual function that does the work:
        const result = await runBenchmarks(github, context);
        // { markdown: string, summary?: string }

        if (!result) {
            throw Error('No benchmark results.');
        }

        if (token) {
            await postOrUpdateComment(github, context, `
                🚀 Benchmark results for ${context.payload.pull_request.head.sha.substring(0,7)}:
                ${result.markdown}
                <sub>(${new Date().toUTCString()})</sub>
            `.trim().replace(/^\s+/gm, ''));
        }

        await finish({
            conclusion: 'success',
            output: {
                title: `Benchmark Results`,
                summary: result.summary
            }
        });
    } catch (e) {
        setFailed(e.message);

        await finish({
            conclusion: 'failure',
            output: {
                title: 'Benchmarks failed',
                summary: `Error: ${e.message}`
            }
        });
    }
})();
```

**Comment by andrewiggins** [comment]
Some questions about how this behavior should work:

1. Should each benchmark post it's own comment or should all benchmarks share and update the same comment (in a non destructive way)? 

    Right now leaning towards all benchmarks sharing the same comment to reduce email noise when PRs are first opened (I assume each comment that gets created sends an email). Do updates to comments send an email?

2. Should benchmarks only run if tests pass?

**Comment by developit** [comment]
Updates to comments don't send an email, so I'm also a fan of that approach. If we wanted to avoid it sending an email at all, the status check gets its own page with arbitrary Markdown, and can also bubble up a simple stats line as the description shown when clicking on a check's green checkmark. I believe checks also post properly from fork branches, which is nice.

I do think benches should only run if tests pass, yeah. When tests fail we can't really trust the benches, so they become additional knock-on failures that don't add their own info.

**Comment by marvinhagemeister** [comment]
1. +1 on reusing the same comment
2. +1 on only running benchmarks when tests pass :+1:

**Comment by raj-al-ghul** [comment]
@andrewiggins thanks for your helper functions, I was attempting to do something similar in one of my own repos and it was great to see a reference.

FYI you can use a markdown comment to mark the comment the bot creates instead of visible text. (just lends itself to a slightly cleaner look).

Also, I'd imagine you'd want to run the benchmark on push and not just create? in which case `context.issue.number` seems to be undefined.

**Comment by andrewiggins** [comment]
Thanks @raj-al-ghul! Actually, the credit goes to @developit for the sample code above :)

Yup, a markdown comment works as well. I personally prefer the visible text though as it provides a way for people who want to do something similar in their PRs a link to the action they can use.

Yea, that matches what I saw too - push doesn't have a `context.issue.number`. We are running our benchmarks on the `pull_request` event which I believe by default [runs when a PR is opened, synchronized, or reopened](https://docs.github.com/en/actions/reference/events-that-trigger-workflows#pull_request) which seems to run on more than just PR create. It seems to always have a `context.issue.number`.

Closing this since I finished v1 of [tachometer-reporter-action](https://github.com/andrewiggins/tachometer-reporter-action) which does what this issue was about.
